gadopt/stokes_integrators.py

```diff
@@ -352,67 +430,153 @@ def set_solver_options(self) -> None:
         if solver_preset is not None:
             match solver_preset:
                 case "direct":
-                    self.solver_parameters.update(direct_stokes_solver_parameters)
+                    self.solver_parameters |= direct_stokes_solver_parameters
                 case "iterative":
-                    self.solver_parameters.update(iterative_stokes_solver_parameters)
+                    self.solver_parameters |= iterative_stokes_solver_parameters
                 case _:
                     raise ValueError("Solver type must be 'direct' or 'iterative'.")
         elif self.mesh.topological_dimension() == 2 and self.mesh.cartesian:
-            self.solver_parameters.update(direct_stokes_solver_parameters)
+            self.solver_parameters |= direct_stokes_solver_parameters
         else:
-            self.solver_parameters.update(iterative_stokes_solver_parameters)
+            self.solver_parameters |= iterative_stokes_solver_parameters
 
-        # Extra options for iterative solvers
+        # Extra monitoring options for iterative solvers
         if self.solver_parameters.get("pc_type") == "fieldsplit":
             if DEBUG >= log_level:
                 self.solver_parameters["fieldsplit_0"]["ksp_converged_reason"] = None
                 self.solver_parameters["fieldsplit_1"]["ksp_monitor"] = None
             elif INFO >= log_level:
                 self.solver_parameters["fieldsplit_1"]["ksp_converged_reason"] = None
 
-                if self.free_surface_map:
-                    # Gather pressure and free surface fields for Schur complement solve
-                    fields_ind = ",".join(map(str, range(1, len(self.solution_split))))
-                    self.solver_parameters.update(
-                        {
-                            "pc_fieldsplit_0_fields": "0",
-                            "pc_fieldsplit_1_fields": fields_ind,
-                        }
-                    )
-                    # Update mass inverse preconditioner
-                    self.solver_parameters["fieldsplit_1"].update(
-                        {"pc_python_type": "gadopt.FreeSurfaceMassInvPC"}
-                    )
+        if self.solver_parameters_update is not None:
+            for key, value in self.solver_parameters_update.items():
+                if isinstance(value, dict):
+                    self.solver_parameters[key] |= value
+                else:
+                    self.solver_parameters[key] = value
+
+    def set_solver(self) -> None:
+        """Sets up the Firedrake variational problem and solver."""
+        if self.constant_jacobian:
```

Comment:
Given that `constant_jacobian` can be `False` in `LinearVariationalProblem`, is it reasonable to use the value of `self.constant_jacobian` as a condition to determine between linear and non-linear problems?

Reply:
But we still wanted a linear solver in that case, right?

Reply:
yeah we still want a linear solver